controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+					return reconcile.Result{}, err
+				}
+			}
+		}
+
+		if vgrContentObj.Spec.VolumeGroupReplicationHandle == "" {
+			logger.Info("Either volumegroupreplicationcontent is not yet created or it is still grouping the volumes to be replicated")
+			return reconcile.Result{}, nil
+		} else {
+			// Create VolumeReplication CR for the current VGR CR, if doesn't exists
+			vrName := fmt.Sprintf("vr-%s", instance.UID)
+			apiGroup := "replication.storage.openshift.io"
+			vrObj = &replicationv1alpha1.VolumeReplication{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      vrName,
+					Namespace: instance.Namespace,
+					Annotations: map[string]string{
+						"replication.storage.openshift.io/volumegroupref": fmt.Sprintf("%s/%s", instance.Name, instance.Namespace),
+					},
+				},
+				Spec: replicationv1alpha1.VolumeReplicationSpec{
+					VolumeReplicationClass: instance.Spec.VolumeReplicationClassName,
+					ReplicationState:       instance.Spec.ReplicationState,
+					DataSource: corev1.TypedLocalObjectReference{
+						APIGroup: &apiGroup,
+						Kind:     instance.Kind,
+						Name:     instance.Name,
+					},
+					AutoResync: instance.Spec.AutoResync,
+				},
+			}
+			_ = controllerutil.SetOwnerReference(instance, vrObj, r.Scheme)
+			foundVRObj := &replicationv1alpha1.VolumeReplication{}
+			err = r.Client.Get(ctx, types.NamespacedName{Name: vrName, Namespace: instance.Namespace}, foundVRObj)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					logger.Info("volumeReplication CR not found for volumeGroupReplication. Creating new volumeReplication resource", "VRName", vrName)
+					err = r.Client.Create(ctx, vrObj)
+					if err != nil {
+						logger.Error(err, "failed to create volumeReplication CR", "VRName", vrName)
+						err := r.setGroupReplicationFailure(instance, logger, err)
+						return reconcile.Result{}, err
+					}
+				} else {
+					logger.Error(err, "failed to get volumeReplication resource", "VRName", vrName)
+					err := r.setGroupReplicationFailure(instance, logger, err)
+					return reconcile.Result{}, err
+				}
+			} else {
+				if !reflect.DeepEqual(foundVRObj, vrObj) {
+					foundVRObj = vrObj
+					err = r.Client.Update(ctx, foundVRObj)
+					if err != nil {
+						logger.Error(err, "failed to update volumeReplication resource", "VRName", vrName)
+						err := r.setGroupReplicationFailure(instance, logger, err)
+						return reconcile.Result{}, err
+					}
+				}
+			}
+
+			// Update the VGR with VolumeReplication resource name, if not present
+			if instance.Spec.VolumeReplicationName == "" {
+				instance.Spec.VolumeReplicationName = vrObj.Name
+				err = r.Client.Update(ctx, instance)
+				if err != nil {
+					logger.Error(err, "failed to update volumeGroupReplication instance", "VGRName", instance.Name)
+					err := r.setGroupReplicationFailure(instance, logger, err)
+					return reconcile.Result{}, err
+				}
+			}
+		}
+	} else {
+		if instance.Status.PersistentVolumeClaimsRefList != nil {
```

Comment:
Lets remove the VGR and VR once they are gone remove the finalizers and annotation on the PVC?

Reply:
We are removing the PVC owner annotation below at line: 332, and later on removing the VR as well at line: 344.